src/libcore/pin.rs

```diff
@@ -1,160 +1,284 @@
+    #[unstable(feature = "pin", issue = "49150")]
+    pub unsafe fn new_unchecked(pointer: P) -> Pin<P> {
+        Pin { pointer }
     }
 
-    /// Get a mutable reference to the data inside of this `PinMut`.
+
+    /// Get a pinned shared reference from this pinned pointer.
     #[unstable(feature = "pin", issue = "49150")]
-    pub fn get_mut(this: PinMut<'a, T>) -> &'a mut T {
-        this.inner
+    pub fn as_ref(this: &Pin<P>) -> Pin<&T> {
+        unsafe { Pin::new_unchecked(&**this) }
     }
 }
 
+impl<P, T> Pin<P> where
+    P: DerefMut<Target = T>,
+    T: ?Sized,
+{
+    /// Get a pinned mutable reference from this pinned pointer.
+    #[unstable(feature = "pin", issue = "49150")]
+    pub fn as_mut(this: &mut Pin<P>) -> Pin<&mut T> {
+        unsafe { Pin::new_unchecked(&mut *this.pointer) }
+    }
 
-#[unstable(feature = "pin", issue = "49150")]
-impl<'a, T: ?Sized> PinMut<'a, T> {
-    /// Construct a new `PinMut` around a reference to some data of a type that
-    /// may or may not implement `Unpin`.
-    ///
-    /// This constructor is unsafe because we do not know what will happen with
-    /// that data after the lifetime of the reference ends. If you cannot guarantee that the
-    /// data will never move again, calling this constructor is invalid.
+    /// Assign a new value to the memory behind the pinned reference.
     #[unstable(feature = "pin", issue = "49150")]
-    pub unsafe fn new_unchecked(reference: &'a mut T) -> PinMut<'a, T> {
-        PinMut { inner: reference }
+    pub fn set(this: Pin<&mut T>, value: T)
+        where T: Sized,
+    {
+        *this.pointer = value;
     }
+}
 
-    /// Reborrow a `PinMut` for a shorter lifetime.
+impl<'a, T> Pin<&'a T> {
+    /// Construct a new pin by mapping the interior value.
+    ///
+    /// For example, if you  wanted to get a `Pin` of a field of something,
+    /// you could use this to get access to that field in one line of code.
     ///
-    /// For example, `PinMut::get_mut(x.reborrow())` (unsafely) returns a
-    /// short-lived mutable reference reborrowing from `x`.
+    /// # Safety
+    ///
+    /// This function is unsafe. You must guarantee that the data you return
+    /// will not move so long as the argument value does not move (for example,
+    /// because it is one of the fields of that value), and also that you do
+    /// not move out of the argument you receive to the interior function.
+    #[unstable(feature = "pin", issue = "49150")]
+    pub unsafe fn map_unchecked<U, F>(this: Pin<&'a T>, func: F) -> Pin<&'a U> where
+        F: FnOnce(&T) -> &U,
+    {
+        let pointer = &*this.pointer;
+        let new_pointer = func(pointer);
+        Pin::new_unchecked(new_pointer)
+    }
+
+    /// Get a safe reference out of a pin.
     #[unstable(feature = "pin", issue = "49150")]
-    pub fn reborrow<'b>(&'b mut self) -> PinMut<'b, T> {
-        PinMut { inner: self.inner }
+    pub fn get(this: Pin<&'a T>) -> &'a T {
```

Comment:
Why do we have `set` for all `Pin<P> where P: DerefMut<Target = T>`, but `get` only for `Pin<&T>`?

Reply:
You can't use 'Deref' here since you want the lifetime to be equal to that of the original reference, not of your borrow of it.

Reply:
Oh, I see... this is actually the method witnessing that we can safely have shared references to pinned data!  It has such an innocent name, I did not realize that. :)